### PR TITLE
auth: also catch end_of_buffer while parsing text input

### PR DESCRIPTION
**Pull request by wjwithagen**
[On FreeBSD]
in KeyRing::decode() text parsing in only executed if decode throws.
This results in an ABORT signal in bin/ceph when running vstart.sh
```
12: Populating config ...
12: /usr/home/Ceph/master/ceph/src/vstart.sh: line 1154: 70347 Done                    cat <<EOF
12: [global]
12: osd_pool_default_size = $OSD_POOL_DEFAULT_SIZE
12: osd_pool_default_min_size = 1
12: mon_pg_warn_min_per_osd = 3
...
...
12: [mgr]
12: mgr/telemetry/nag = false
12: mgr/telemetry/enable = false
12:
12: EOF
12:
12:      70348 Abort trap              (core dumped) | $CEPH_BIN/ceph -c $conf_fn config assimilate-conf -i -
```

Only buffer::error is caught in `KeyRing.cc`, but reading config text sometimes
lets the buffer_length run into absurd numbers. This results in a `copy()` call with too large length.
And this generates `end_of_buffer`.

So we need to catch that as well.

However adding buffer::end_of_buffer as exception in another case
does not seem to work, because it is not matched????
So catching all exceptions, and continue to perhaps trap in  decode_plaintext()

fixes: https://tracker.ceph.com/issues/41717

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/auth/KeyRing.cc [hunk 1]
```diff
@@ -203,6 +203,12 @@ void KeyRing::decode(bufferlist::const_iterator& bl) {
   } catch (buffer::error& err) {
     keys.clear();
     decode_plaintext(start_pos);
+  }catch (buffer::end_of_buffer& err) {
```
Comment (liewegas) [inline src/auth/KeyRing.cc]: nit: add space between } and catch?

Comment (tchaikov) [reply src/auth/KeyRing.cc]: i don't understand why shall we catch `buffer::end_of_buffer` if we already catch `buffer::error`. as the former is derived from the latter.

Comment (tchaikov) [reply src/auth/KeyRing.cc]: reading the description of the tracker ticket

> Getting out of bounds access in the buffer, generating a trap.

accessing out of array bounds is undefined behavior. how can we catch it? what exactly exception is thrown?

**Comment by wjwithagen** [comment]
@tchaikov
I'll change the commit text, and I'll see if I can still dig out the actual error printed.
It is just ABORT and a core-dump.

As for using `end_of_buffer`, your knowledge could help.

I expected the code to actually trigger the `catch(buffer::error)` as you suggest, but it does not.
So I used the third catch to actually find out what the exception was, and it printed `end_of_buffer`.
But the weird thing is, that ATM not even that `catch()` is working as expecteed.
And we still end up in the third `catch()` where I have removed the debugging code to print `e.what()`

So I guess that in the case of FreeBSD, Clang, boost-1.70 that assumption `end_of_buffer` in `buffer::error` does not hold???

**Comment by tchaikov** [comment]
@wjwithagen i need to understand the reason why clang behaves like this. and FWIW, i ran into the same issue when testing on the MacOS build of cephfs client.

**Comment by wjwithagen** [comment]
@tchaikov 
Could be a clang issue? 
FreeBSD current has now native Clang 9.0 as compiler. But I have not yet upgrade my dev platform to see if that fixes it. Normally I've got a lot of work to make everything work again which such major OS upgrades.
But I guess it could be worth it.

I'll also ask one oth the Clang compiler porters on FreeBSD on his opinion, how knows, we might get lucky.

**Comment by tchaikov** [comment]
@wjwithagen JFYI, i compiled Ceph with clang-10 on GNU/Linux, and it runs just fine. can we wait until FreeBSD includes llvm110 port, and see if it addresses your issue?

**Comment by wjwithagen** [comment]
@tchaikov 
FreeBSD has only recently upgraded its compiler to Clang 9.0
Even ports does not have anything above 9.0
So there is not going to be much of a chance that 10.0 will be readily available as native compiler on 12.x of 13.x...

On the other hand, submitting this PR does not change much (anything) to the functionallity of the already correct working programs. So there should not be a problem in commiting the PR for those.
What I can do is start generating warnings when OS=FreeBSD and CLANG_VERSION >= 10.0.
That'll tigger me (or others to remove the "workaround")
The ultimate sollution is to do what I do now: I take it along in the Ports patchset. But then that misses out on platforms with Clang 9.0 that are not FreeBSD.
So I'm very much in favour of merging.

**Comment by wjwithagen** [comment]
unstale ping

**Comment by tchaikov** [comment]
after a second thought, i wonder if it's related to symbol visibility of `ceph::buffer::error` and its derived classes. could you try out following patch?

```diff
diff --git a/src/common/error_code.h b/src/common/error_code.h
index 67c6cb448d4..5205525c3c3 100644
--- a/src/common/error_code.h
+++ b/src/common/error_code.h
@@ -101,7 +101,7 @@ struct is_error_condition_enum<::ceph::buffer::errc> {
 };
 }

-namespace ceph::buffer {
+namespace ceph::buffer __attribute__((visibility("default"))) {
 inline namespace v15_2_0 {

 // implicit conversion:
```

**Comment by wjwithagen** [comment]
@tchaikov 
This does not work for clang 9, the compiler that comes with FreeBSD 12.1.
Then I tried with clang10 and clang-devel which will become clang11
Same ABORT

So that does not fixes the problem here.
Which is sort of wierd given that the ABORT is only the consequence of the catch not doing its work.
And that is just the compiler?? Or could the linker also be involved in this.

For the time I'm just keeping this code for patching on FreeBSD.
Perhaps we just include this exclusive for FreeBSD, patch has been standing here long enough??

**Comment by wjwithagen** [comment]
ping alive

**Pull request closed without merging**